File: test/metabolism/demo/tmp/11_paper_draft_report/scripts/convert_html_to_pdf.R
#将HTML报告转换为PDF
html_path <- "G:/OmicsWorks/test/metabolism/demo/analysis/report.html"
pdf_path <- "G:/OmicsWorks/test/metabolism/demo/analysis/report.pdf"

#检查HTML文件是否存在
if (!file.exists(html_path)) {
 stop(paste("HTML文件不存在:", html_path))
}

cat(paste("HTML文件大小:", file.info(html_path)$size, "字节\n"))

#尝试直接调用wkhtmltopdf
cmd <- sprintf(
 'wkhtmltopdf --page-size A3 --margin-top15mm --margin-bottom15mm --margin-left15mm --margin-right15mm --enable-local-file-access --encoding UTF-8 "%s" "%s"',
 html_path, pdf_path
)

cat(paste("执行命令:\n", cmd, "\n\n"))

#使用system执行
exit_code <- system(cmd, intern = FALSE, ignore.stdout = FALSE, ignore.stderr = FALSE, wait = TRUE)
cat(paste("退出码:", exit_code, "\n"))

#检查PDF是否生成
if (file.exists(pdf_path)) {
 cat(paste("\nPDF成功生成:", pdf_path, "\n"))
 cat(paste("PDF文件大小:", file.info(pdf_path)$size, "字节\n"))
} else {
 cat("\nPDF生成失败，尝试其他方法...\n")
 
 #尝试用R的webshot包
 if (require(webshot, quietly = TRUE)) {
 cat("使用webshot包...\n")
 webshot::webshot(url = html_path, file = pdf_path, vwidth =1400, vheight =900)
 } else {
 cat("webshot包不可用\n")
 }
 
 if (file.exists(pdf_path)) {
 cat(paste("\nPDF成功生成:", pdf_path, "\n"))
 cat(paste("PDF文件大小:", file.info(pdf_path)$size, "字节\n"))
 } else {
 cat("所有方法均失败\n")
 }
}

File: test/metabolism/demo/tmp/11_paper_draft_report/scripts/convert_html_to_pdf.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""将HTML报告转换为PDF"""

import subprocess
import sys
import os

html_path = r"G:/OmicsWorks/test/metabolism/demo/analysis/report.html"
pdf_path = r"G:/OmicsWorks/test/metabolism/demo/analysis/report.pdf"

#检查HTML文件是否存在
if not os.path.exists(html_path):
 print(f"错误: HTML文件不存在: {html_path}")
 sys.exit(1)

print(f"HTML文件大小: {os.path.getsize(html_path)}字节")

#尝试多种可能的wkhtmltopdf路径
possible_paths = [
 "wkhtmltopdf",
 r"C:\Program Files\wkhtmltopdf\bin\wkhtmltopdf.exe",
 r"C:\Program Files (x86)\wkhtmltopdf\bin\wkhtmltopdf.exe",
]

wkhtmltopdf_cmd = None
for cmd in possible_paths:
 try:
 result = subprocess.run([cmd, "--version"], capture_output=True, text=True, timeout=10)
 if result.returncode ==0:
 wkhtmltopdf_cmd = cmd
 print(f"找到 wkhtmltopdf: {cmd}")
 print(f"版本信息: {result.stdout.strip()[:100]}")
 break
 except (FileNotFoundError, subprocess.TimeoutExpired):
 continue

if wkhtmltopdf_cmd is None:
 print("未找到 wkhtmltopdf，尝试使用 Python pdfkit库...")
 try:
 import pdfkit
 print("pdfkit库可用，尝试转换...")
 options = {
 'page-size': 'A3',
 'margin-top': '15mm',
 'margin-bottom': '15mm',
 'margin-left': '15mm',
 'margin-right': '15mm',
 'enable-local-file-access': '',
 'encoding': 'UTF-8',
 }
 pdfkit.from_file(html_path, pdf_path, options=options)
 print(f"PDF已生成: {pdf_path}")
 print(f"PDF文件大小: {os.path.getsize(pdf_path)}字节")
 sys.exit(0)
 except ImportError:
 print("pdfkit库不可用")
 print("请安装: pip install pdfkit")
 print("或安装 wkhtmltopdf: https://wkhtmltopdf.org/downloads.html")
 sys.exit(1)

#使用wkhtmltopdf命令行转换
cmd_args = [
 wkhtmltopdf_cmd,
 "--page-size", "A3",
 "--margin-top", "15mm",
 "--margin-bottom", "15mm",
 "--margin-left", "15mm",
 "--margin-right", "15mm",
 "--enable-local-file-access",
 "--encoding", "UTF-8",
 html_path,
 pdf_path
]

print(f"执行命令: {' '.join(cmd_args)}")
result = subprocess.run(cmd_args, capture_output=True, text=True, timeout=120)

print(f"返回码: {result.returncode}")
if result.stdout:
 print(f"标准输出: {result.stdout[:500]}")
if result.stderr:
 print(f"标准错误: {result.stderr[:500]}")

if result.returncode ==0 and os.path.exists(pdf_path):
 print(f"PDF成功生成: {pdf_path}")
 print(f"PDF文件大小: {os.path.getsize(pdf_path)}字节")
else:
 print("PDF生成失败")
 sys.exit(1)
